torch/utils/data/datapipes/iter/tararchivereader.py

```diff
@@ -52,15 +50,13 @@ def __iter__(self) -> Iterator[Tuple[str, BufferedIOBase]]:
                         warnings.warn("failed to extract file {} from source tarfile {}".format(tarinfo.name, pathname))
                         raise tarfile.ExtractError
                     inner_pathname = os.path.normpath(os.path.join(pathname, tarinfo.name))
-                    # Add a reference of the source tarfile into extracted_fobj, so the source
-                    # tarfile handle won't be released until all the extracted file objs are destroyed.
-                    extracted_fobj.source_ref = tar  # type: ignore[attr-defined]
-                    # typing.cast is used here to silence mypy's type checker
-                    yield (inner_pathname, cast(BufferedIOBase, extracted_fobj))
+                    yield (inner_pathname, extracted_fobj)  # type: ignore[misc]
             except Exception as e:
                 warnings.warn(
                     "Unable to extract files from corrupted tarfile stream {} due to: {}, abort!".format(pathname, e))
                 raise e
+            finally:
+                data_stream.close()
```

Comment:
Try to understand the level to put finally. It seems we will call finally multiple times in the iterator. Is this the right way?
In my opinion, finally represents the destructor for the iterator object. Then, multiple finally doesn't make sense to me.

If you want to close every single `data_stream`, you can directly call `data_stream.close()` without finally. And, if the iterator has not been depleted and then destructed, we should have the finally closure to close the last `data_stream`

Reply:
I would suggest you to do the same thing as we discussed by adding an outer try-finally

Reply:
I think we both agree that each `data_stream` from the `DataPipe` should be closed. The simplest way would be, as you suggested, adding `data_stream.close()` at the end of the each loop after the `yield`. The issue is that if the instance prematurely ends (due to exception, reset, etc), that statement will not be reached.

By adding `finally`, it ensures that the statement be executed under all circumstances here (after each iteration and right before the instance is destroyed), because a `finally` clause is always executed.

That is my rationale behind adding `finally` there rather than having two `try` statements.

Reply:
I think the difference is that I see `finally` is not only executed as a destructor, it is always executed even when the object is not deconstructed.

Reply:
I see. Then, can we verify the close is always executed within finally clause?

Reply:
I verified by adding these print statements to the function to test:
```
            except Exception as e:
                warnings.warn(
                    "Unable to extract files from corrupted tarfile stream {} due to: {}, abort!".format(pathname, e))
                raise e
            finally:
                print("Executing CLOSE")
                data_stream.close()
            print("END OF LOOP")
```
For the last test case (where the iterator was reset before completion) it prints the following:

```
Executing CLOSE  # An iterator completes and the data_stream closes
END OF LOOP # From when the iterator is exhausted
Executing CLOSE. # First data_stream is closed later when the object is deleted
```

The result is as expected because "END OF LOOP" is only reached once when the iterator is exhausted, whereas both `data_stream` are closed.


Do you have a suggestion on how to add a test case? Or is the example above suffice? The issue is that we need to access a variable inside the function that is not an output.

Reply:
Then, if nothing breaks, the last data_stream will be closed twice? Am I right?

Reply:
No, just once when the object is deconstructed. Before that, the object will be in a state after `yield` but before `finally`.

Reply:
Sounds reasonable. Then it's good to go